### PR TITLE
[cpu] implement flash attention backward

### PR DESCRIPTION
**Pull request by Valentine233**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105131
* #107128
* #104863
* __->__ #104693
* #103826
* #104584
* #104583

Feature RFC: https://github.com/pytorch/rfcs/pull/56.

The flash attention CPU kernel is added, for backward path FP32. Parallelization is on the dimensions of batch size and head number.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by mingfeima** [comment]
you may cherry pick https://github.com/pytorch/pytorch/pull/104583 and https://github.com/pytorch/pytorch/pull/104584 to this stack and replace `_mkl_gemm` with `cpublas::gemm`, then the mkl platform specific CI fails will pass.

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 1]
```diff
@@ -370,6 +370,231 @@ void cpu_flash_attention(
...
+  int64_t kStrideH = key.stride(2);
+  int64_t vStrideB = value.stride(0);
+  int64_t vStrideN = value.stride(1);
+  int64_t vStrideH = value.stride(2);
+  int64_t oStrideB = out.stride(0);
+  int64_t oStrideM = out.stride(1);
+  int64_t oStrideH = out.stride(2);
+  int64_t lStrideB = logsumexp.stride(0);
+  int64_t lStrideM = logsumexp.stride(1);
+  int64_t lStrideH = logsumexp.stride(2);
```
Comment (jgong5) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:423]: Can we simplify the data ptr access with `TensorAccessor`? We don't have to retrieve all these strides vars and also it is simpler to get the data ptr with the index variables. Not sure about the perf impact though.

Comment (Valentine233) [reply aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:423]: Good idea, thanks! As the change is relatively big and the perf impact is also needed to verify, is it OK to do this change in another separate PR?

Comment (jgong5) [reply aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:423]: Sure.

**Review by jgong5 — APPROVED** [review]

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 2]
```diff
@@ -403,8 +628,51 @@ void flash_attention_kernel_impl(
...
+
+  AT_DISPATCH_SWITCH(query.scalar_type(), "flash_attention_backward",
+    AT_DISPATCH_CASE(ScalarType::Float, [&] {
+      cpu_flash_attention_backward<scalar_t, 128, 256>(
+          grad_q, grad_k, grad_v, grad_out_contig,
+          query, key, value, out, logsumexp,
+          cum_seq_q, cum_seq_k, max_q, max_k, dropout_p,
+          is_causal, philox_seed, philox_offset, scale);
+    });
+    // AT_DISPATCH_CASE(ScalarType::BFloat16, [&] {
```
Comment (drisspg) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:663]: Do we still need this?

**Review by drisspg — APPROVED** [review]

**Pull request closed without merging**